### PR TITLE
[Enhancement] Improve the randomness of the random function

### LINKED ISSUE #34026: Some issues with StarRocks being compatible with Trino
## Enhancement

**Test version: StarRocks 3.1.0**

### 1. `rand` function
The `rand` function has two problems:
1. `select * from xxx where rand() < 0.1;` We tested this query and either returned empty or all data. We expected to return about 1/10 of the data. Inconsistent with Trino behavior;
2. `rand(n)`, when there are parameters, SR represents seed, and Trino represents the output range. The semantics of the two are different;

### 2. `regexp_extract` function
In StarRocks, the `regexp_extract` function returns an empty string when there is no matching content. Trino returns NULL. The two behave inconsistently.

I hope to get everyone’s discussion, feedback and help.

### PR DESCRIPTION
**Pull request by zenoyang**
Why I'm doing:
The original `random` function of SR was implemented through the `rand_r` function. The randomness was very poor and usually did not meet expectations.

before:
image

after:
image

What I'm doing:

Fixes #34026

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by imay** [comment]
@zenoyang 
After run is slower than current, I think you pasted the wrong picture.

File: be/src/exprs/math_functions.cpp [hunk 1]
```diff
@@ -687,18 +687,11 @@ StatusOr<ColumnPtr> MathFunctions::conv_string(FunctionContext* context, const C
...
-    }
-    return seed;
-}
-
+std::uniform_real_distribution<double>* MathFunctions::_distribution =
+        new std::uniform_real_distribution<double>(lower_bound, upper_bound);
 Status MathFunctions::rand_prepare(FunctionContext* context, FunctionContext::FunctionStateScope scope) {
     if (scope == FunctionContext::THREAD_LOCAL) {
-        int64_t seed = 0;
+        std::mt19937_64* generator = new std::mt19937_64();
```
Comment (stdpain) [inline be/src/exprs/math_functions.cpp:694]: replace with static thread_local std::mt19937 generator;

Comment (stdpain) [reply be/src/exprs/math_functions.cpp:694]: here was my test result after this change:
image

File: be/src/exprs/math_functions.h [hunk 2]
```diff
@@ -523,6 +524,9 @@ class MathFunctions {
     static const int32_t MIN_BASE = 2;
     static const int32_t MAX_BASE = 36;
     static const char* _s_alphanumeric_chars;
+    constexpr static const double lower_bound = 0.0;
+    constexpr static const double upper_bound = 1.0;
+    static std::uniform_real_distribution<double>* _distribution;
```
Comment (stdpain) [inline be/src/exprs/math_functions.h:529]: static variable is enough

File: be/src/exprs/math_functions.cpp [hunk 3]
```diff
@@ -715,27 +708,32 @@ Status MathFunctions::rand_prepare(FunctionContext* context, FunctionContext::Fu
             }
 
             int64_t seed_value = ColumnHelper::get_const_value<TYPE_BIGINT>(seed_column);
-            seed = seed_value;
+            generator->seed(seed_value);
         } else {
-            seed = GetCurrentTimeNanos();
+            generator->seed(std::random_device()());
```
Comment (stdpain) [inline be/src/exprs/math_functions.cpp:713]: remove it

File: be/src/exprs/math_functions.cpp [hunk 4]
```diff
@@ -715,27 +708,32 @@ Status MathFunctions::rand_prepare(FunctionContext* context, FunctionContext::Fu
...
+        auto state = reinterpret_cast<std::mt19937_64*>(context->get_function_state(FunctionContext::THREAD_LOCAL));
+        delete state;
+    }
     return Status::OK();
 }
 
 StatusOr<ColumnPtr> MathFunctions::rand(FunctionContext* context, const Columns& columns) {
-    int32_t num_rows = ColumnHelper::get_const_value<TYPE_INT>(columns[columns.size() - 1]);
-    void* state = context->get_function_state(FunctionContext::THREAD_LOCAL);
+    std::mt19937_64* generator =
```
Comment (stdpain) [inline be/src/exprs/math_functions.cpp:729]: declare static thread local here

File: be/src/exprs/math_functions.cpp (hunk 1 above)
Comment (zenoyang) [reply be/src/exprs/math_functions.cpp:694]: Thanks, I'll try to modify and test it

**Comment by zenoyang** [comment]
> @zenoyang After run is slower than current, I think you pasted the wrong picture.

No, that is the result of the intermediate state of development. The performance was not very good at the beginning.

**Comment by zenoyang** [comment]
> @zenoyang After run is slower than current, I think you pasted the wrong picture.

No, that is the result of the intermediate state of development. The performance was not very good at the beginning.

File: be/src/exprs/math_functions.cpp (hunk 4 above)
Comment (zenoyang) [reply be/src/exprs/math_functions.cpp]: Thanks, I modified it based on your suggestions, and the test performance is in line with expectations. Please help to review it again. @stdpain

**Review by stdpain — DISMISSED** [review]

**Comment by stdpain** [comment]
CI clang-format failed in 
```patch
--- /home/runner/_work/starrocks/starrocks/be/src/exprs/math_functions.cpp
+++ /home/runner/_work/starrocks/starrocks/be/src/exprs/math_functions.cpp (after clang format)
/home/runner/_work/starrocks/starrocks/be/src/exprs/math_functions.cpp had clang-format style issues
@@ -34,7 +34,7 @@
 static const double MAX_EXP_PARAMETER = std::log(std::numeric_limits<double>::max());
 
 static std::uniform_real_distribution<double> distribution(0.0, 1.0);
-static thread_local std::mt1[9](https://github.com/StarRocks/starrocks/actions/runs/6928566833/job/18844627873?pr=35199#step:6:10)937_64 generator { std::random_device{}() };
+static thread_local std::mt[19](https://github.com/StarRocks/starrocks/actions/runs/6928566833/job/18844627873?pr=35199#step:6:20)937_64 generator{std::random_device{}()};
 
 // ==== basic check rules =========
 DEFINE_UNARY_FN_WITH_IMPL(NegativeCheck, value) {
```

**Comment by stdpain** [comment]
cla also need check

**Comment by stdpain** [comment]
sign off also need check. try to execute
```
git commit -s -m "[Enhancement] Improv...."
```

**Comment by zenoyang** [comment]
> sign off also need check. try to execute签核也需要检查。尝试执行
> 
> ```
> git commit -s -m "[Enhancement] Improv...."
> ```

done

**Review by stdpain — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**